File: Array/Minimise_the_Heights_II.py
def getMinDiff(arr, k):
    arr.sort()
    res = arr[len(arr) - 1] - arr[0]
    small = arr[0] + k
    large = arr[len(arr) - 1] - k
    minE = 0
    maxE = 0

    for i in range(0, len(arr)-1):
        minE = min(small,  arr[i+1]-k)
        maxE = max(large, arr[i]+k)
        if minE < 0:
            continue
        res = min(res, maxE-minE)

    return res 

if __name__ == '__main__':
    print(getMinDiff([1, 5, 8, 10], 2))
    print(getMinDiff([3, 9, 12, 16, 20], 3))

File: Array/Minimise_the_Heights_II.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> getArray()
{
    int n;
    cout << "Enter size of array: ";
    cin >> n;
    vector<int> arr(n);
    cout << "Enter the array:";
    for (int i = 0; i < n; i++)
    {
        cin >> arr[i];
    }

    return arr;
}

void printArray(vector<int> arr, int size)
{
    for (int i = 0; i < size; i++)
    {
        cout << arr[i] << " ";
    }

    cout << endl;
}

int getMinDiff(vector<int> arr, int k)
{
    sort(arr.begin(), arr.end());
    int res = arr[arr.size() - 1] - arr[0];
    int small = arr[0] + k;
    int large = arr[arr.size() - 1] - k;
    int minE, maxE;

    for (int i = 0; i < arr.size(); i++)
    {
        minE = min(small, arr[i + 1] - k);
        maxE = max(large, arr[i] + k);
        if (minE < 0)
            continue;
        res = min(res, maxE - minE);
    }
    return res;
}

int main()
{
    vector<int> arr;
    arr = getArray();

    int k;
    cout << "Enter value of k: ";
    cin >> k;

    cout << "Original array: " << endl;
    printArray(arr, arr.size());

    int minDiff = getMinDiff(arr, k);

    cout << "Min max height: " << minDiff;

    return 0;
}